nxdialog/nxdialog.py

```diff
@@ -0,0 +1,368 @@
+
+import optparse
+import os
+import signal
+import sys
+
+import gtk
+
+import pygtk
+pygtk.require("2.0")
+
+PROGRAM = "nxdialog"
+
+DISCONNECT = 1
+TERMINATE = 2
+
+EXIT_SUCCESS = 0
+EXIT_FAILURE = 1
+
+CANCEL_TEXT = "Cancel"
+DISCONNECT_TEXT = "Disconnect"
+TERMINATE_TEXT = "Terminate"
+YES_TEXT = "Yes"
+NO_TEXT = "No"
+
+DLG_TYPE_ERROR = "error"
+DLG_TYPE_OK = "ok"
+DLG_TYPE_PANIC = "panic"
+DLG_TYPE_PULLDOWN = "pulldown"
+DLG_TYPE_QUIT = "quit"
+DLG_TYPE_YESNO = "yesno"
+DLG_TYPE_YESNOSUSPEND = "yesnosuspend"
+
+VALID_DLG_TYPES = frozenset([
+    DLG_TYPE_ERROR,
+    DLG_TYPE_OK,
+    DLG_TYPE_PANIC,
+    DLG_TYPE_PULLDOWN,
+    DLG_TYPE_QUIT,
+    DLG_TYPE_YESNO,
+    DLG_TYPE_YESNOSUSPEND,
+    ])
+
+class PullDownMenu(object):
+  """ Shows a popup menu to disconnect/terminate session. """
+
+  def __init__(self, window_id):
+    """ Initializes this class.
+
+    @type window_id: int
+    @param window_id: X11 window id of target window
+
+    """
+    self._window_id = window_id
+    self._result = None
+
+  def Show(self):
+    """ Shows popup and returns result. """
+
+    win = gtk.gdk.window_foreign_new(self._window_id)
+
+    menu = gtk.Menu()
+    menu.connect("deactivate", self._MenuDeactivate)
+
+    # TODO: Show title item in bold font
+    title = gtk.MenuItem(label="Session control")
+    title.set_sensitive(False)
+    menu.append(title)
+
+    disconnect = gtk.MenuItem(label=DISCONNECT_TEXT)
+    disconnect.connect("activate", self._ItemActivate, DISCONNECT)
+    menu.append(disconnect)
+
+    terminate = gtk.MenuItem(label=TERMINATE_TEXT)
+    terminate.connect("activate", self._ItemActivate, TERMINATE)
+    menu.append(terminate)
+
+    menu.append(gtk.SeparatorMenuItem())
+
+    cancel = gtk.MenuItem(label=CANCEL_TEXT)
+    menu.append(cancel)
+
+    menu.show_all()
+
+    menu.popup(parent_menu_shell=None, parent_menu_item=None,
+               func=self._PosMenu, data=win,
+               button=0, activate_time=gtk.get_current_event_time())
+
+    gtk.main()
+
+    return self._result
+
+  def _ItemActivate(self, _, result):
+    """ called when a menu item is selected """
+    self._result = result
+    gtk.main_quit()
+
+  def _MenuDeactivate(self, _):
+    """ called when menu is deactivated """
+    gtk.main_quit()
+
+  def _PosMenu(self, menu, parent):
+    """ Positions menu at the top center of the parent window. """
+    # Get parent geometry and origin
+    (_, _, win_width, _, _) = parent.get_geometry()
+    (win_x, win_y) = parent.get_origin()
+
+    # Calculate width of menu
+    (menu_width, _) = menu.size_request()
+
+    # Calculate center
+    x = win_x + ((win_width - menu_width) / 2)
+
+    return (x, win_y, True)
+
+
+def ShowYesNoSuspendBox(title, text):
+  """ Shows a message box to disconnect/terminate session.
+
+  @type title: str
+  @param title: Message box title
+  @type text: str
+  @param text: Message box text
+  @return: Choosen action
+
```

Comment:
This is a little standalone tool. No need for an api documentation. I'd prefer to leave the Epydoc parts just because they are easily understandable. No one will ever generate an api document from that.